Use enums/booleans for enumerated attributes.

aptls opened the pull request:
Modify the attribute getters/setters for attributes defined in
the DTD with enumerated values. Use enums for most string values
and boolean where the values in the DTD are "true|false".

The behavior on the getters is currently to return the value of
the member variable, returning null when the value is unset
(false for boolean attributes). This is different from the
generated methods, which returned the default attribute value
as specified in the DTD even when the member variable was unset. The intent
is to model what will be present in the generated XML.

Add tests that verify that the expected enum/boolean is returned
on unmarshal and that the setters result in the correct string in
the XML.

Commit:
Use enums/booleans for enumerated attributes.

Modify the attribute getters/setters for attributes defined in
the DTD with enumerated values. Use enums for most string values
and boolean where the values in the DTD are "true|false".

The behavior on the getters is currently to return the value of
the member variable, returning null when the value is unset
(false for boolean attributes). This is different from the
generated methods, which returned the default attribute value
as specified in the DTD even when the member variable was unset. The intent
is to model what will be present in the generated XML.

Add tests that verify that the expected enum/boolean is returned
on unmarshal and that the setters result in the correct string in
the XML.

wiarlawd commented on src/com/google/enterprise/gsafeed/Acl.java:
```diff
@@ -47,6 +48,41 @@
+    }
+
+    public static InheritanceType fromString(String value) {
+      if (value.equals("child-overrides")) {
```
Maybe:

    for (InheritanceType type : InheritanceType.values) {
      if (type.xmlValue.equals(value)) {
        return type;
      }
    }
    throw new IllegalArgumentException(value);

wiarlawd commented on src/com/google/enterprise/gsafeed/Acl.java:
```diff
@@ -70,24 +106,45 @@ public Acl setUrl(String value) {
+      return null;
     } else {
-      return inheritanceType;
+      return InheritanceType.fromString(inheritanceType);
```
This seems awkward. Can we marshal the enum values directly, rather than store them as strings. This post at least suggests that we can:
http://blog.bdoughan.com/2011/08/jaxb-and-enums.html

wiarlawd commented on src/com/google/enterprise/gsafeed/Acl.java:
```diff
@@ -70,24 +106,45 @@ public Acl setUrl(String value) {
    */
   public Acl setInheritanceType(String value) {
-    this.inheritanceType = value;
+    if (Strings.isNullOrEmpty(value)) {
```
What's the purpose of allowing empty strings to act like nulls?

Maybe, if we don't allow empty strings, and we keep the string representation of enums, we could move the null check into fromString and remove it from the getter and String setter.

wiarlawd commented on src/com/google/enterprise/gsafeed/Metadata.java:
```diff
@@ -43,14 +43,16 @@
-  public String getOverwriteAcls() {
+  public boolean getOverwriteAcls() {
+    // The default as defined in the dtd is true. Return that
+    // here to reflect the underlying meaning, but don't set the
```
This comment seems odd in the getter. What does it mean to not the set string here? Also, you're not allowing setOverwriteAcls(null) to reset the default, like you do for the enums. Maybe at least allow that for setOverwriteAcls(String), and maybe think about using Boolean instead of boolean, and live with the autoboxing?

wiarlawd commented on src/com/google/enterprise/gsafeed/Record.java:
```diff
@@ -306,27 +460,63 @@ public String getCrawlOnce() {
+  /**
+   * Sets the value of the scoring property.
+   *
+   * @param value allowed object is {@link Scorng} or null
```
s/Scorng/Scoring/

wiarlawd commented on test/com/google/enterprise/gsafeed/AclTest.java:
```diff
@@ -41,4 +50,107 @@ public void testAcl() throws Exception {
+
+  @Test
+  public void testGetInheritanceTypeInvalid() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
```
If this exception isn't thrown until the getter is called, then move these down to make that more clear (and not spuriously pass the test if unmarshal throws an IllegalArgumentException). Otherwise, remove the call to the getter.

wiarlawd commented on test/com/google/enterprise/gsafeed/AclTest.java:
```diff
@@ -14,16 +14,25 @@
  * Test Acl.
  */
 public class AclTest {
+  private static final Charset UTF_8 = Charset.forName("UTF-8");
```
These aren't used. Maybe you added JaxbUtil later on?

wiarlawd commented on src/com/google/enterprise/gsafeed/Acl.java:
```diff
@@ -70,24 +106,45 @@ public Acl setUrl(String value) {
+   * @return this object
+   */
+  public Acl setInheritanceType(InheritanceType value) {
+    if (value == null) {
```
If we keep this string representation, this code block is pretty repetitive. Maybe:

    this.inheritanceType = (value == null) ? null : value.toString();
or

    this.inheritanceType = toString(value, null);

where toString is an implementation of Objects.toString (which we can't use; it's Java 7).

wiarlawd reviewed (commented):
Minor stuff, questions, and suggestions.

Commit:
Code review changes.

Remove string setters. Use Enum/Boolean in the member field
definitions instead of preserving the original string field
types. Remove unused code in tests.

wiarlawd commented on test/com/google/enterprise/gsafeed/AclTest.java:
```diff
@@ -83,65 +80,53 @@ public void testGetInheritanceTypeNotSet() throws Exception {
-    thrown.expectMessage("foo");
     Acl acl = unmarshal("<acl inheritance-type='foo'/>");
-    acl.getInheritanceType();
+    assertEquals(null, acl.getInheritanceType());
```
This is the price we pay for unmarshalling to an enum? Invalid values are just silently ignored? What happens with an invalid value if we unmarshal the full XML using the DTD?

aptls replied:
This is what we get without validation. With validation, the parser throws an exception.

wiarlawd commented on test/com/google/enterprise/gsafeed/AclTest.java:
```diff
@@ -83,65 +80,53 @@ public void testGetInheritanceTypeNotSet() throws Exception {
     thrown.expect(IllegalArgumentException.class);
     thrown.expectMessage("foo");
-    new Acl().setInheritanceType("foo");
+    Acl.InheritanceType.fromString("foo");
```
Maybe add a test for a valid use of fromString?

wiarlawd commented on src/com/google/enterprise/gsafeed/Acl.java:
```diff
@@ -40,13 +41,45 @@
+    }
+
+    public static InheritanceType fromString(String value) {
+      for (InheritanceType inheritanceType : InheritanceType.values()) {
```
In looking at this, we get fromString(null) -> IllegalArgumentException(null). Do we want either fromString(null) -> null (since that's a valid argument to the setters, and it avoids callers who want to use fromString from making the check themselves), or fromString(null) -> NPE?

wiarlawd reviewed (commented):
This looks really clean.

Commit:
Code review changes.

Add tests of fromString with the enum values. Let fromString return
null when passed null. Add string methods to GsafeedHelper and
XmlgroupsHelper. Add tests in RecordTest of invalid attribute values
for enum and boolean attributes.

wiarlawd commented on src/com/google/enterprise/gsafeed/Content.java:
```diff
@@ -35,27 +34,59 @@
+  @XmlEnum(String.class)
+  public static enum Encoding {
+    @XmlEnumValue("base64binary")
+    BASE_64_BINARY("base64binary"),
```
Base64 seems like one word to me, so I think these (and the one for meta-encoding) should be BASE64.... I also wonder about having a fixed rule for the strings, like the value.replace("-", "_").toUpperCase(US) we talked about, so that this would be BASE64BINARY by virtue of the unhyphenated string value. Not convinced either way.

wiarlawd replied:
No one else seems to have an opinion, so just s/BASE_64/BASE64/g.

wiarlawd reviewed (approved):
LGTM++ with a waffle.

Please wait for PJ's approval as well.

PapaJoltOmega commented on build.xml:
```diff
@@ -32,6 +32,9 @@
 
   <path id="test.classpath">
+    <fileset dir="${third_party.dir}">
+        <include name="guava/*.jar"/>
```
where are these coming from? they're not in this CL?

wiarlawd replied:
Took me a little while to figure that out as well. The "tests", such as they were, did not exercise enough prod code to hit the prod dependency on Guava. We could instead include gsa.build.classpath here, which might be slightly more conventional.

wiarlawd replied:
I was mistaken about this. The initial commit was using Strings.isNullOrEmpty, but that's been removed, and this can go away. The only use of Guava in gsafeed is an `@VisibleForTesting` in the copied GsaFeedFileSender.

PapaJoltOmega reviewed (approved):
Thank you

wiarlawd reviewed (approved):
LGTM++, resolving issues.

Commit:
Code review changes.
